src/main/java/org/folio/circulation/domain/OverduePeriodCalculator.java

```diff
@@ -0,0 +1,125 @@
+  }
+
+  private static int getGracePeriodMinutes(Loan loan) {
+    LoanPolicy loanPolicy = loan.getLoanPolicy();
+    if (loanPolicy == null) {
+      return ZERO_MINUTES;
+    }
+
+    int duration = loanPolicy.getGracePeriodDuration();
+    switch (loanPolicy.getGracePeriodInterval()) {
+    case MONTHS:
+      return duration * MINUTES_PER_MONTH;
+    case WEEKS:
+      return duration * MINUTES_PER_WEEK;
+    case DAYS:
+      return duration * MINUTES_PER_DAY;
+    case HOURS:
+      return duration * MINUTES_PER_HOUR;
+    case INCORRECT:
+      return ZERO_MINUTES;
+    case MINUTES:
+    default:
+      return duration;
+    }
+  }
+
+  private static CompletableFuture<Result<Integer>> getOverdueMinutes(Loan loan, DateTime systemTime, Clients clients) {
+    OverdueFinePolicy overdueFinePolicy = loan.getOverdueFinePolicy();
+    if (overdueFinePolicy == null || overdueFinePolicy.getCountClosed() == null) {
+      return completedFuture(succeeded(ZERO_MINUTES));
+    }
+
+    DateTime dueDate = loan.getDueDate();
+    boolean shouldCountClosed = overdueFinePolicy.getCountClosed();
+    if (shouldCountClosed) {
+      int overdueMinutes = minutesBetween(dueDate, systemTime).getMinutes();
+      return completedFuture(succeeded(overdueMinutes));
+    } else {
+      return new CalendarRepository(clients)
+        .fetchOpeningPeriodsBetweenDates(loan.getCheckoutServicePointId(), dueDate, systemTime, false)
+      .thenApply(r -> r.next(OverduePeriodCalculator::getOpeningDaysDurationMinutes));
+    }
+  }
+
+  private static Result<Integer> getOpeningDaysDurationMinutes(List<OpeningPeriod> openingDays) {
+    return succeeded(
+      openingDays.stream()
+        .map(OpeningPeriod::getOpeningDay)
+        .mapToInt(day -> day.getAllDay() ? MINUTES_PER_DAY : getOpeningDayDurationMinutes(day))
+        .sum()
+    );
+  }
+
+  private static int getOpeningDayDurationMinutes(OpeningDay openingDay) {
+    return openingDay.getOpeningHour()
+      .stream()
+      .mapToInt(OverduePeriodCalculator::getOpeningHourDurationMinutes)
+      .sum();
+  }
+
+  private static int getOpeningHourDurationMinutes(OpeningHour openingHour) {
+    LocalTime startTime = openingHour.getStartTime();
+    LocalTime endTime = openingHour.getEndTime();
+    if (ObjectUtils.allNotNull(startTime, endTime) && endTime.isAfter(startTime)) {
+      return getMinutesOfDay(endTime) - getMinutesOfDay(startTime);
+    }
+    return ZERO_MINUTES;
+  }
+
+  private static int getMinutesOfDay(LocalTime time) {
+    return time.getHourOfDay() * MINUTES_PER_HOUR + time.getMinuteOfHour();
+  }
+
+  private static Result<Integer> adjustOverdueWithGracePeriod(Loan loan, int overdueMinutes, int gracePeriodMinutes) {
+    int result = shouldIgnoreGracePeriod(loan)
+      ? overdueMinutes
+      : Math.max(overdueMinutes - gracePeriodMinutes, ZERO_MINUTES);
+
+    return Result.succeeded(result);
+  }
+
+  private static boolean shouldIgnoreGracePeriod(Loan loan) {
```

Comment:
![image](https://user-images.githubusercontent.com/9094650/72796790-a1e9fe00-3c48-11ea-8e74-9b1c743110b1.png)

It seems some important checks have not been covered.

Reply:
covered